eras/alonzo/impl/src/Cardano/Ledger/Alonzo/UTxO.hs

```diff
@@ -53,3 +74,76 @@ getInputDataHashesTxBody (UTxO mp) txBody hashScriptMap =
             -- the Alonzo era already set the precedent with datum hashes, and several dapp
             -- developers see this as a helpful feature.
             _ -> ans
+
+-- |
+-- Uses of inputs in ‘txscripts’ and ‘neededScripts’
+-- There are currently 3 sets of inputs (spending, collateral, reference). A particular TxInput
+-- can appear in more than one of the sets. Even in all three at the same, but that may not be
+-- a really useful case. Inputs are where you find scripts with the 'Spending' purpose.
+--
+-- 1) Collateral inputs are only spent if phase two fails. Their corresponding TxOut can only have
+--    Key (not Script) Pay credentials, so ‘neededScripts’ does not look there.
+-- 2) Reference inputs are not spent in the current Tx, unless that same input also appears in one
+--    of the other sets. If that is not the case, their credentials are never needed, so anyone can
+--    access the inline datums and scripts in their corresponding TxOut, without needing any
+--    authorizing credentials. So ‘neededScripts’ does not look there.
+-- 3) Spending inputs are always spent. So their Pay credentials are always needed.
+--
+-- Collect information (purpose and ScriptHash) about all the Credentials that refer to scripts
+-- that will be needed to run in a TxBody in the Utxow rule. Note there may be credentials that
+-- cannot be run, so are not collected. In Babbage, reference inputs, fit that description.
+-- Purposes include
+-- 1) Spending (payment script credentials, but NOT staking scripts) in the Addr of a TxOut, pointed
+--    to by some input that needs authorization. Be sure (getField @"inputs" txb) gets all such inputs.
+--    In some Eras there may be multiple sets of inputs, which ones should be included? Currently that
+--    is only the spending inputs. Because collateral inputs can only have key-locked credentials,
+--    and reference inputs are never authorized. That might not always be the case.
+-- 2) Rewarding (Withdrawals),
+-- 3) Minting (minted field), and
+-- 4) Certifying (Delegating) scripts.
+--
+-- 'getAlonzoScriptsNeeded' is an aggregation of the needed Credentials referring to
+-- Scripts used in Utxow rule.  The flip side of 'getAlonzoScriptsNeeded' (which collects
+-- script hashes) is 'txscripts' which finds the actual scripts. We maintain an invariant
+-- that every script credential refers to some actual script.  This is tested in the test
+-- function 'validateMissingScripts' in the Utxow rule.
```

Comment:
thanks for documenting this so well!

Reply:
Oh, no. That's not me. I just merely moved the haddock over from another module :smile: